### PR TITLE
Cross library build for Spark 2/3

### LINKED ISSUE #407: [Announcement]: master branch will move to support Spark 3.0.
The `master` branch is planned to move to have a dependency on Spark 3.0 after Hyperspace 0.5 is released; this indicates that Spark 2.4 will be supported by Hyperspace version 0.5.x. `branch-0.5` will be created at that time, and any critical bugs will be back-ported and new patch versions (e.g., 0.5.1, 0.5.2) will be released if necessary. However, there is no plan to back-port any new features.

Please leave comments if you have any concerns.

### PR DESCRIPTION
**Pull request by clee704**
Fixes #407

### What is the context for this pull request?

 - **Tracking Issue**: #407 
 - **Parent Issue**: N/A
 - **Dependencies**: N/A

### What changes were proposed in this pull request?

Support for Spark 3 in addition to Spark 2. The build produces three JARs for the following combinations:

- Spark 3.0 (3.0.1) / Scala 2.12
- Spark 2.4 (2.4.2) / Scala 2.12
- Spark 2.4 (2.4.2) / Scala 2.11

Note that Spark 3 doesn't support Scala 2.11.

### Does this PR introduce _any_ user-facing change?

The artifact name is changed from `hyperspace-core_<scala version>-<hyperspace version>.jar` to `hyperspace-core_spark<spark version>_<scala version>-<hyperspace version>.jar`, where `<spark version>` is either `2.4` or `3.0`.

### How was this patch tested?

Existing unit tests

### CODE REVIEW

**Comment by imback82** [comment]
🚀

**Comment by clee704** [comment]
Although there are still failing tests, those are tests checking the actual shapes of the plans, which changed from 2.4 to 3.0. I'll keep working on them but other than that this PR is ready.

**Comment by clee704** [comment]
TODO: package python files. sbt-spark-package was removed due to incompatibility with sbt 1.5.0. The only job of the plugin seemed to be packaging python files. Please let me know if it's not the case.

File: src/main/scala/com/microsoft/hyperspace/actions/RefreshActionBase.scala [hunk 1]
```diff
@@ -71,16 +71,18 @@ private[actions] abstract class RefreshActionBase(
...
+    val latestRelation = relationMetadata.refresh()
     val dataSchema = DataType.fromJson(latestRelation.dataSchemaJson).asInstanceOf[StructType]
-    val df = spark.read
-      .schema(dataSchema)
+    var df = spark.read
       .format(latestRelation.fileFormat)
       .options(latestRelation.options)
+    if (relationMetadata.supportsUserSpecifiedSchema) {
+      df = df.schema(dataSchema)
+    }
```
Comment (sezruby) [inline src/main/scala/com/microsoft/hyperspace/actions/RefreshActionBase.scala:85]: Is this better?
```suggestion
    val df = {
      if (relationMetadata.supportsUserSpecifiedSchema) spark.read.schema(dataSchema)
      else spark.read
    }.format(latestRelation.fileFormat)
      .options(latestRelation.options)
```

File: src/main/scala/com/microsoft/hyperspace/index/sources/default/DefaultFileBasedRelationMetadata.scala [hunk 2]
```diff
@@ -36,4 +36,6 @@ class DefaultFileBasedRelationMetadata(metadata: Relation) extends FileBasedRela
   override def enrichIndexProperties(properties: Map[String, String]): Map[String, String] = {
     properties
   }
+
+  override def supportsUserSpecifiedSchema: Boolean = true
```
Comment (sezruby) [inline src/main/scala/com/microsoft/hyperspace/index/sources/default/DefaultFileBasedRelationMetadata.scala:40]: nit: for consistency
```suggestion
  override def supportUserSpecifiedSchema: Boolean = true
```

File: src/main/scala/com/microsoft/hyperspace/index/sources/iceberg/IcebergRelation.scala [hunk 3]
```diff
@@ -153,19 +146,17 @@ class IcebergRelation(spark: SparkSession, override val plan: DataSourceV2Relati
   /**
    * Options of the current relation.
    */
-  override def options: Map[String, String] = plan.options
+  override def options: Map[String, String] = Map[String, String]()
```
Comment (sezruby) [inline src/main/scala/com/microsoft/hyperspace/index/sources/iceberg/IcebergRelation.scala:149]: Could you add some comment for this?

File: src/test/scala/com/microsoft/hyperspace/index/HybridScanSuite.scala [hunk 4]
```diff
@@ -396,32 +396,37 @@ trait HybridScanSuite extends QueryTest with HyperspaceSuite {
...
-        withSQLConf(
-          SQLConf.WHOLESTAGE_CODEGEN_ENABLED.key -> "false",
-          IndexConstants.INDEX_HYBRID_SCAN_ENABLED -> "false") {
-          val join = joinQuery()
-          checkAnswer(join, baseQuery)
-          assert(basePlan.equals(join.queryExecution.optimizedPlan))
-        }
+      // Dynamic pruning creates a dynamic filter, with different ids every time
+      // thus making basePlan.equals(join) fail
+      withSQLConf("spark.sql.optimizer.dynamicPartitionPruning.enabled" -> "false") {
```
Comment (sezruby) [inline src/test/scala/com/microsoft/hyperspace/index/HybridScanSuite.scala:402]: 👍 

You can change the config in `beforeAll` or `before` in HybridScanForPartitionedDataTest

File: src/test/scala/com/microsoft/hyperspace/index/IndexManagerTest.scala [hunk 5]
```diff
@@ -342,7 +341,7 @@ class IndexManagerTest extends HyperspaceSuite with SQLHelper {
     // 3. Call optimize. Check the metadata. It should not contain small index files created
     // during refresh operations.
     withTempPathAsString { testPath =>
-      withSQLConf(OPTIMIZE_FILE_SIZE_THRESHOLD -> "900") {
+      withSQLConf(OPTIMIZE_FILE_SIZE_THRESHOLD -> "910") {
```
Comment (sezruby) [inline src/test/scala/com/microsoft/hyperspace/index/IndexManagerTest.scala:344]: is this change necessary?

File: build.sbt [hunk 6]
```diff
@@ -14,84 +14,106 @@
  * limitations under the License.
  */
 
-name := "hyperspace-core"
+import Dependencies._
 
-sparkVersion := "2.4.2"
+lazy val spark24 = Version(2, 4, 2)
```
Comment (sezruby) [inline build.sbt:19]: Not in this change, but we could update Spark2 version to 2.4.4 or higher one. WDYT? @imback82

File: spark2/src/main/scala-spark2/com/microsoft/hyperspace/index/sources/iceberg/IcebergUtils.scala [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
+import org.apache.spark.sql.execution.datasources.v2.DataSourceV2Relation
+import org.apache.spark.sql.sources.v2.DataSourceOptions
+
+import com.microsoft.hyperspace.util.JavaConverters._
+import com.microsoft.hyperspace.util.SparkUtils
+
+object IcebergUtils {
+
+  def isIcebergRelation(plan: LogicalPlan): Boolean = plan match {
```
Comment (sezruby) [inline spark2/src/main/scala-spark2/com/microsoft/hyperspace/index/sources/iceberg/IcebergUtils.scala:36]: @andrei-ionescu could you review the iceberg related changes?

File: spark2/src/main/scala-spark2/com/microsoft/hyperspace/index/sources/delta/DeltaUtils.scala [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.microsoft.hyperspace.index.sources.delta
+
+import org.apache.spark.sql.delta.actions.AddFile
+import org.apache.spark.sql.delta.files.TahoeLogFileIndex
+
+object DeltaUtils {
```
Comment (sezruby) [inline spark2/src/main/scala-spark2/com/microsoft/hyperspace/index/sources/delta/DeltaUtils.scala:22]: DeltaLakeUtils? for consistency

File: src/test/scala/com/microsoft/hyperspace/index/IndexManagerTest.scala (hunk 5 above)
Comment (clee704) [reply src/test/scala/com/microsoft/hyperspace/index/IndexManagerTest.scala:344]: It happens that the small files are between 900 and 910 bytes in Spark 3 (it was ~890 bytes in Spark 2.4). Large files are like ~950 bytes. This seems arbitrary and there should be a better way, but let's keep it this way in this PR.

File: src/test/scala/com/microsoft/hyperspace/index/HybridScanSuite.scala (hunk 4 above)
Comment (sezruby) [reply src/test/scala/com/microsoft/hyperspace/index/HybridScanSuite.scala:402]: yep seems so as long as there's no exception in after() https://github.com/yyuu/scalatest/blob/master/src/main/scala/org/scalatest/BeforeAndAfter.scala

File: src/main/scala/com/microsoft/hyperspace/actions/RefreshActionBase.scala [hunk 9]
```diff
@@ -71,14 +71,16 @@ private[actions] abstract class RefreshActionBase(
...
+    val relationMetadata = Hyperspace
       .getContext(spark)
       .sourceProviderManager
       .getRelationMetadata(relations.head)
-      .refresh()
+    val latestRelation = relationMetadata.refresh()
     val dataSchema = DataType.fromJson(latestRelation.dataSchemaJson).asInstanceOf[StructType]
-    val df = spark.read
-      .schema(dataSchema)
+    var df = {
```
Comment (sezruby) [inline src/main/scala/com/microsoft/hyperspace/actions/RefreshActionBase.scala:80]: `val` + `}.format(...` in my previous comment is done by scalafmt

Comment (clee704) [reply src/main/scala/com/microsoft/hyperspace/actions/RefreshActionBase.scala:80]: Done

**Comment by clee704** [comment]
Python now works. The remaining is updating expected plans. However, I found that Sort operator is not gone when indices are used for SortMergeJoin. This might take some time...

**Comment by imback82** [comment]
> However, I found that Sort operator is not gone when indices are used for SortMergeJoin. This might take some time...

Is it affected by this (try setting it to true)? https://github.com/apache/spark/blob/e503b9c462676c381ba51ad48edd968d6e184fc2/sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala#L2587

**Comment by sezruby** [comment]
Seems build.sbt doesn't work on Windows + on OS X `testOnly` triggers all tests.

**Comment by clee704** [comment]
> Seems build.sbt doesn't work on Windows + on OS X `testOnly` triggers all tests.

I've uploaded Git aliases and updated README for Windows developers. Please have a look. Although I can compile and package on Windows, tests seem to fail on Windows for a reason which is not related to this PR. Specifically, `Utils.deleteRecursively` seems to fail on Windows.

For OS X, as I work on OS X, it should work. Could you check it out again and please let me know how you can reproduce the issue?

**Comment by sezruby** [comment]
@clee704 Thanks, the provided git alias works on Windows. 
For testOnly, I found out that quotes are required - `sbt "testOnly testName"`

**Comment by clee704** [comment]
Ready to merge

**Review by sezruby — COMMENTED** [review]
LGTM thanks @clee704! BTW Can we disable the plan stability suite for spark 3 and handle it with another PR?

**Review by imback82 — COMMENTED** [review]
> BTW Can we disable the plan stability suite for spark 3 and handle it with another PR?

+1, so we can reference this commit more easily.

File: build.sbt [hunk 10]
```diff
@@ -14,84 +14,106 @@
...
+// add the following source directories for different Spark versions:
+// * scala-spark<spark major version>
+// * scala-spark<spark major version>.<spark minor version>
+// * scala-<scala version>-spark<spark major version>
+// * scala-<scala version>-spark<spark major version>.<spark minor version>
+lazy val addSparkVersionSpecificSourceDirectories = unmanagedSourceDirectories ++= Seq(
+  sourceDirectory.value / s"scala-spark${sparkVersion.value.major}",
+  sourceDirectory.value / s"scala-spark${sparkVersion.value.short}",
+  sourceDirectory.value / s"scala-${scalaBinaryVersion.value}-spark${sparkVersion.value.major}",
+  sourceDirectory.value / s"scala-${scalaBinaryVersion.value}-spark${sparkVersion.value.short}")
```
[the remark this answers is not in the document]
Comment (clee704) [reply build.sbt:70]: These can be used when we want to support different minor versions (e.g. 3.0 and 3.1).

File: azure-pipelines.yml [hunk 11]
```diff
@@ -17,69 +17,87 @@ jobs:
...
+      # If not a pull request, publish artifacts.
+      - ${{ if and(ne(variables['System.TeamProject'], 'public'), notin(variables['Build.Reason'], 'PullRequest')) }}:
+        - script: sbt ++2.12.8 "project spark3" package
+          displayName: 'Running $sbt package'
+        - task: CopyFiles@2
+          displayName: 'Copy hyperspace-core JAR'
+          inputs:
+            sourceFolder: '$(Build.SourcesDirectory)/target/'
+            contents: '**/*.jar'
+            targetFolder: '$(Build.ArtifactStagingDirectory)/hyperspace-core_spark3.0/'
```
[the remark this answers is not in the document]
Comment (clee704) [reply azure-pipelines.yml:95]: Since we are heavily depending on the internal APIs, incompatibilities are very likely. I've already seen some incompatibilities between 3.0 and 3.1. It is our choice - we can either support only the latest minor version (e.g. when we move to 3.1 we drop the support for 3.0), or multiple minor versions.

The current scheme which includes minor version in the artifact name is more suitable if we are going to support multiple minor versions. There weren't huge differences between 2.4 and 3.0, so I think it is feasible to support multiple minor versions.

File: src/main/scala/com/microsoft/hyperspace/index/sources/iceberg/IcebergFileBasedSource.scala [hunk 12]
```diff
@@ -28,8 +27,9 @@ import com.microsoft.hyperspace.index.sources.{FileBasedRelation, FileBasedRelat
  * Iceberg file-based source provider.
  *
  * This source can support relations that meet the following criteria:
- *   - The relation is with [[DataSourceV2Relation]]
- *   - The source of [[DataSourceV2Relation]] is an [[IcebergSource]]
+ *   - The relation is with [[DataSourceV2Relation]] or [[DataSourceV2ScanRelation]]
+ *   - The source of [[DataSourceV2Relation]] is an [[IcebergSource]],
+ *     or the table is an Iceberg table
```
[the remark this answers is not in the document]
Comment (clee704) [reply src/main/scala/com/microsoft/hyperspace/index/sources/iceberg/IcebergFileBasedSource.scala:32]: Done

File: src/main/scala/org/apache/spark/deploy/hyperspace/SparkHadoopUtil.scala [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.deploy.hyperspace
+
+import org.apache.hadoop.fs.{FileSystem, Path}
+
+// SparkHadoopUtil became package-private in Spark 3
```
[the remark this answers is not in the document]
Comment (clee704) [reply src/main/scala/org/apache/spark/deploy/hyperspace/SparkHadoopUtil.scala]: Done

File: azure-pipelines.yml [hunk 14]
```diff
@@ -17,69 +17,87 @@ jobs:
...
         - task: PublishBuildArtifacts@1
           displayName: 'Publish Hyperspace artifacts'
           inputs:
-            artifactName: 'hyperspace-core'
-            pathtoPublish: '$(Build.ArtifactStagingDirectory)/hyperspace-core/'
+            artifactName: 'hyperspace-core_spark2.4'
+            pathtoPublish: '$(Build.ArtifactStagingDirectory)/hyperspace-core_spark2.4/'
+
+  - job: Build_Spark3_2_12
+    displayName: 'Build sources and run unit tests for Spark 3.0 / Scala 2.12'
```
[the remark this answers is not in the document]
Comment (clee704) [reply azure-pipelines.yml:74]: Spark 3.1 won't work with our JAR built for 3.0 due to incompatible changes in the API we rely on.

File: azure-pipelines.yml (hunk 11 above)
[the remark this answers is not in the document]
Comment (imback82) [reply azure-pipelines.yml:95]: Sounds good and we should support multiple minor versions. Let's get started on Spark 3.1 support once this PR is merged.

**Comment by imback82** [comment]
I will play around with this for one more day and merge if there are no other comments.

FYI @sezruby @apoorvedave1 @thugsatbay @andrei-ionescu @dmytroDragan (the current PRs may not merge cleanly).

File: azure-pipelines.yml (hunk 14 above)
[the remark this answers is not in the document]
Comment (imback82) [reply azure-pipelines.yml:74]: Sorry, I meant supporting 3.1.

File: spark3/src/main/scala-spark3/com/microsoft/hyperspace/util/SparkUtils.scala [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+
+package com.microsoft.hyperspace.util
+
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.sql.catalyst.expressions.Expression
+import org.apache.spark.sql.catalyst.plans.JoinType
+import org.apache.spark.sql.catalyst.plans.logical.{Join, JoinHint, LogicalPlan}
+import org.apache.spark.sql.execution.{QueryExecution, SQLExecution}
+
+object SparkUtils {
```
[the remark this answers is not in the document]
Comment (clee704) [reply spark3/src/main/scala-spark3/com/microsoft/hyperspace/util/SparkUtils.scala]: Done

File: azure-pipelines.yml (hunk 11 above)
[the remark this answers is not in the document]
Comment (clee704) [reply azure-pipelines.yml:95]: Okay, then I'll make this PR more friendly to the future 3.1 support.

File: README.md [hunk 16]
```diff
@@ -21,6 +21,20 @@ provided by the bot. You will only need to do this once across all repos using o
...
+
+```sh
+$ git config --local include.path ../dev/.gitconfig
+$ git replace-symlinks # replace symlinks with junctions
+$ git restore-symlinks # restore symlinks
+```
+
+### Using IntelliJ
+
+You can use the built-in sbt shell in IntelliJ without any problems. However, the built-in "Build Project" command may not work. To fix the issue, go to Project Structure -> Project Settings -> Modules, remove the "root" module, and mark `src/main/scala` as Sources for the "spark2_4" and "spark3_0" modules, as well as `src/main/scala-spark2` and `src/main/scala-spark3` directories for corresponding modules.
```
Comment (imback82) [inline README.md:36]: I also had to do the following to get these compiled in IntelliJ from a fresh clone:
* Change from `Zinc` to `IDEA` under `Build,Execution,Deployment` -> `Compiler` -> `Scala Compiler` ->` Incrementality type`.
* remove the `hyperspace-sources` module

Can you also mention marking `src/test/scala` (and others) as `Tests` so that you can run them easily inside IntelliJ?

Comment (imback82) [reply README.md:36]: One more thing is that `BuildInfo` plugin doesn't seem to run if I build it from IDE, so I had to compile from the CLI `sbt` first, then build it from IDE. This could be an existing behavior, but we may want to call it out while we are at it.

File: src/main/scala/com/microsoft/hyperspace/actions/RefreshActionBase.scala [hunk 17]
```diff
@@ -71,15 +71,16 @@ private[actions] abstract class RefreshActionBase(
...
       .sourceProviderManager
       .getRelationMetadata(relations.head)
-      .refresh()
+    val latestRelation = relationMetadata.refresh()
     val dataSchema = DataType.fromJson(latestRelation.dataSchemaJson).asInstanceOf[StructType]
-    val df = spark.read
-      .schema(dataSchema)
-      .format(latestRelation.fileFormat)
+    val df = {
+      if (relationMetadata.canSupportUserSpecifiedSchema) spark.read.schema(dataSchema)
```
Comment (imback82) [inline src/main/scala/com/microsoft/hyperspace/actions/RefreshActionBase.scala:81]: oh, is this fixing an existing bug?

**Review by imback82 — COMMENTED** [review]
LGTM, thanks @clee704! This PR is 🚀!

**Review by imback82 — APPROVED** [review]

**Comment by imback82** [comment]
Merging to master! Let's address pending comments as a separate PR.

File: src/main/scala/com/microsoft/hyperspace/actions/RefreshActionBase.scala (hunk 17 above)
Comment (clee704) [reply src/main/scala/com/microsoft/hyperspace/actions/RefreshActionBase.scala:81]: Maybe they were ignored in Spark 2. In Spark 3 it throws an exception if schema is specified for Delta or Iceberg relations.

File: README.md (hunk 16 above)
Comment (clee704) [reply README.md:36]: I didn't have change that, maybe it's different by OS. I'll update the instruction.

Comment (clee704) [reply README.md:36]: BuildInfo was already not working in IntelliJ before this PR. Anyway, let's mention it too.
